File: scripts/keep-alive.py
#!/usr/bin/env python3

import os
import time
import requests
from datetime import datetime
import signal
import sys

# Configuration
WEBSITE_URL = os.getenv('WEBSITE_URL', 'https://your-app-name.onrender.com')
PING_INTERVAL = int(os.getenv('PING_INTERVAL', '14'))  # minutes
HEALTH_ENDPOINT = '/api/classes'  # Use existing API endpoint

class KeepAlive:
    def __init__(self):
        self.running = True
        signal.signal(signal.SIGINT, self.shutdown)
        signal.signal(signal.SIGTERM, self.shutdown)
    
    def shutdown(self, signum, frame):
        print("\n🛑 Keep-alive stopped")
        self.running = False
        sys.exit(0)
    
    def ping_website(self):
        url = f"{WEBSITE_URL}{HEALTH_ENDPOINT}"
        timestamp = datetime.now().isoformat()
        
        try:
            start_time = time.time()
            response = requests.get(url, timeout=30)
            duration = int((time.time() - start_time) * 1000)
            
            if 200 <= response.status_code < 300:
                print(f"✅ {timestamp} - Ping successful ({response.status_code}) - {duration}ms")
            else:
                print(f"⚠️ {timestamp} - Ping returned {response.status_code} - {duration}ms")
                
        except requests.exceptions.Timeout:
            print(f"⏰ {timestamp} - Ping timeout after 30 seconds")
        except requests.exceptions.RequestException as e:
            print(f"❌ {timestamp} - Ping failed: {str(e)}")
    
    def start(self):
        print(f"🚀 Starting keep-alive for {WEBSITE_URL}")
        print(f"📡 Pinging every {PING_INTERVAL} minutes")
        print(f"🎯 Health endpoint: {HEALTH_ENDPOINT}")
        print("---")
        
        # Initial ping
        self.ping_website()
        
        # Main loop
        while self.running:
            time.sleep(PING_INTERVAL * 60)  # Convert minutes to seconds
            if self.running:
                self.ping_website()

if __name__ == "__main__":
    keep_alive = KeepAlive()
    keep_alive.start()

File: scripts/keep-alive.js
#!/usr/bin/env node

import https from 'https';
import http from 'http';

// Configuration
const WEBSITE_URL = process.env.WEBSITE_URL || 'https://your-app-name.onrender.com';
const PING_INTERVAL = parseInt(process.env.PING_INTERVAL) || 14; // minutes
const HEALTH_ENDPOINT = '/api/classes'; // Use existing API endpoint

function pingWebsite() {
  const url = new URL(WEBSITE_URL + HEALTH_ENDPOINT);
  const client = url.protocol === 'https:' ? https : http;
  
  const startTime = Date.now();
  
  const req = client.get(url, (res) => {
    const duration = Date.now() - startTime;
    const timestamp = new Date().toISOString();
    
    if (res.statusCode >= 200 && res.statusCode < 300) {
      console.log(`✅ ${timestamp} - Ping successful (${res.statusCode}) - ${duration}ms`);
    } else {
      console.log(`⚠️ ${timestamp} - Ping returned ${res.statusCode} - ${duration}ms`);
    }
  });
  
  req.on('error', (err) => {
    const timestamp = new Date().toISOString();
    console.log(`❌ ${timestamp} - Ping failed: ${err.message}`);
  });
  
  req.setTimeout(30000, () => {
    req.destroy();
    const timestamp = new Date().toISOString();
    console.log(`⏰ ${timestamp} - Ping timeout after 30 seconds`);
  });
}

function startKeepAlive() {
  console.log(`🚀 Starting keep-alive for ${WEBSITE_URL}`);
  console.log(`📡 Pinging every ${PING_INTERVAL} minutes`);
  console.log(`🎯 Health endpoint: ${HEALTH_ENDPOINT}`);
  console.log('---');
  
  // Initial ping
  pingWebsite();
  
  // Set up interval
  setInterval(pingWebsite, PING_INTERVAL * 60 * 1000);
}

// Handle graceful shutdown
process.on('SIGINT', () => {
  console.log('\n🛑 Keep-alive stopped');
  process.exit(0);
});

process.on('SIGTERM', () => {
  console.log('\n🛑 Keep-alive terminated');
  process.exit(0);
});

// Start the keep-alive service
startKeepAlive();